pandas/tests/internals/test_internals.py

```diff
@@ -1242,3 +1242,34 @@ def test_interleave_non_unique_cols():
     assert df_unique.values.shape == df.values.shape
     tm.assert_numpy_array_equal(df_unique.values[0], df.values[0])
     tm.assert_numpy_array_equal(df_unique.values[1], df.values[1])
+
+
+def test_assign_df_datetime_column_with_index():
```

Comment:
if you want to write a test that directly tests the Block.setitem behavior that would go here.  These tests are testing `DataFrame.loc.__setitem__`, so go in tests.frame.indexing.test_loc